beeline-core/src/main/java/io/honeycomb/beeline/tracing/ids/W3CTraceIdProvider.java

```diff
@@ -0,0 +1,92 @@
+package io.honeycomb.beeline.tracing.ids;
+
+import java.util.Random;
+
+/**
+ * Creates Trace and Span IDs that conform to the W3C specification.
+ * This is inline with Beeline instrumentations in other programming languages.
+ *
+ * <h1>Thread-safety</h1>
+ * Instances of this class are thread-safe and can be shared.
+ */
+public class W3CTraceIdProvider implements TraceIdProvider {
+    private static final TraceIdProvider INSTANCE = new W3CTraceIdProvider();
+    private static final Random RAND = new Random();
+    private static final int TRACEID_LENGTH = 16;
+    private static final int SPANID_LENGTH = 8;
+
+    public static TraceIdProvider getInstance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public String generateId() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public String generateTraceId() {
+        byte[] bytes = new byte[TRACEID_LENGTH];
+        RAND.nextBytes(bytes);
+        return bytesToHex(bytes);
+    }
+
+    @Override
+    public String generateSpanId() {
+        byte[] bytes = new byte[SPANID_LENGTH];
+        RAND.nextBytes(bytes);
+        return bytesToHex(bytes);
+    }
+
+    // Taken from Stackoverflow post: https://stackoverflow.com/a/9855338/329666
+    private static final char[] HEX_ARRAY = "0123456789abcdef".toCharArray();
```

Comment:
It makes me sad that there aren't system library functions to do a bunch of this encoding and validation.

Reply:
Yeah, me too - there are some external libraries that can do this but I didn't want to take an extra dependency. The performance of this approach is pretty good too.